lib/using_restricted_array.rb

```diff
@@ -6,56 +6,127 @@
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because the time it take to run the program will run depending on how long the array is. 
+# Space complexity: O(n) because the space used will depend on how long the length of the array is. 
 def print_array(array)
-  raise NotImplementedError
+  i = 0 
+  until array[i] == nil
+    print "#{array[i]} "
+    i += 1
+  end 
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because it will run until i is equal to length, which varies depending on array.
+# Space complexity: O(n) because space will be used until i is equal to length.
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0 
+  while i < length
+    if value_to_find == array[i]
+      return true
+    end 
+    i += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) as the time taken will depend on the length of the array
+# Space complexity: O(n) as the space taken will depend on the length of the array
 def find_largest(array, length)
-  raise NotImplementedError
+  if length == 0
+    return nil
+  end 
+
+  largest_value = array[0]
+  i = 1
+  while i < length
+    if array[i] > largest_value
+      largest_value = array[i]
+    end 
+    i = i + 1
+  end
+
+  return largest_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) as the time taken will depend on the length of the array
+# Space complexity: O(n) as the space taken will depend on the length of the array
 def find_smallest(array, length)
-  raise NotImplementedError
+  if length == 0
+    return nil
+  end 
+
+  smallest_value = array[0]
+  i = 1
+  while i < length
+    if array[i] < smallest_value
+      smallest_value = array[i]
+    end 
+    i = i + 1
+  end
+
+  return smallest_value
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) as the time taken will depend on the length of the array
+# Space complexity: O(n) as the space taken will depend on the length of the array
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+  j = length - 1
+
+  while i < j
+    temp = array[i]
+    array[i] = array[j]
+    array[j] = temp
+
+    i += 1
+    j -= 1
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) as the space taken will depend on the length of the array although faster as it cuts the time in half as a binary search
```

Comment:
Since with each iteration it cuts the number of elements to check in half, this has time complexity O(log n). 

Since the data being used does not change no matter how long the array is this is O(1).